### PR TITLE
Move macro translation logic into MacroTranslator class

### PR DESCRIPTION
**Pull request by SilverMight**
Break up the logic of macro translations and move them from analyze transformations into MacroTranslator which handles taking a MacroMap and emitting all the translations.

This will make it easier to have a general MacroTranslationStats class as an instance member that can handle recording translation stats that we need and also just translation logic easier to understand.

Note that I'd like to make this project an actual package w/ sub-packages, and consider moving `run_maki_on_compile_commands.py` into a util script for Maki as it's entirely separate from the rest of MerC, but I'm more focused on getting evaluation done currently.

Tested and verified same translations were emitted on bash.

### CODE REVIEW

File: macrotranslator.py
Comment (PappasBrent) [inline macrotranslator.py:14]: I think it's more idiomatic to replace all these `| None` types with [`Optional`](https://docs.python.org/3/library/typing.html#typing.Optional)s

```suggestion
                                    mm: MacroMap) -> dict[Macro, Optional[str]]:
```

You'll need to add `from typing import Optional` to the top of the file as well.

File: macrotranslator.py
Comment (PappasBrent) [inline macrotranslator.py:15]:
```suggestion
        translationMap: dict[Macro, Optional[str]] = {}
```

You get the idea; not commenting on the later occurrences but you can change those as well.

File: macrotranslator.py
Comment (PappasBrent) [inline macrotranslator.py]: Ideally any of the logic for checking if we should translate a macro should be in the `ie_def()` function in `predicates/interface_equivalent.py` so that we keep it all in one place (also makes it easier to write about our translation rules in the paper to have them all in one place as well).

Comment (PappasBrent) [reply macrotranslator.py]: You don't have to move this function's logic to the `ie_def()` necessarily; I'm just asking that all the logic for checking if a macro is translatable be in one function/method for concision.

File: macrotranslator.py
Comment (PappasBrent) [inline macrotranslator.py]: I think you can use Maki's `IsExpansionTypeVoid` property here to perform this check more robustly and without needing to use regex.

**Review by PappasBrent — CHANGES_REQUESTED** [review]

File: macrotranslator.py
Comment (SilverMight) [reply macrotranslator.py]: > Ideally any of the logic for checking if we should translate a macro should be in the `ie_def()` function in `predicates/interface_equivalent.py` so that we keep it all in one place (also makes it easier to write about our translation rules in the paper to have them all in one place as well).

I more so put skips here because I didn't feel that it fit in ie_def (defining what is and isn't interface equivalent) but was more because of limitations with Maki. I.e we could translate macros containing a function pointer, but Maki doesn't properly emit the type signature. And we could translate macros with a DeclRefExpr, but that's a more complex case to handle.

Do you have any more thoughts on structuring this? I wouldn't mind moving it to ie_def but I'd have to restructure to make tracking the amount of skips due to these two reasons possible.

File: macrotranslator.py
Comment (SilverMight) [reply macrotranslator.py]: Good catch, updated.

File: macrotranslator.py
Comment (SilverMight) [reply macrotranslator.py:14]: AFAIK it's [encouraged](https://discuss.python.org/t/clarification-for-pep-604-is-foo-int-none-to-replace-all-use-of-foo-optional-int/26945/6) to use the "T | None" syntax though Optional isn't deprecated. 

I personally find it a little clearer to understand in this context because it's clear that a macro either has a string translation or None.

Comment (PappasBrent) [reply macrotranslator.py:14]: Ah I wasn't aware of that; interesting. So I suppose `|` is preferable over `Optional` in return types as well argument types? In any case, if you prefer the union syntax then keep it since you are the main developer on the translator at the moment.

File: macrotranslator.py
Comment (PappasBrent) [reply macrotranslator.py]: Ok I see what you mean; separating the theoretical limitations from the technical limitations.
Feel free to leave these checks here then; just add a comment making it clear that the checks here are implementation details preventing translation and not because of semantic issues that would arise from replacing the macro definition with a variable/enum/function.

> Do you have any more thoughts on structuring this?

If you wanted to restructure then what I would do is print the tracking information for each property first, and then perform the translatability check for `should_skip_macro()`, the perform the regular check for `is_def()`. The new definition of `ie_def()` would look something like this:

```python3
def ie_def(m: Macro, pd: PreprocessorData) -> bool:
    is_ = pd.mm[m]
    # We only analyze top-level non-argument invocations
    assert all([i.IsTopLevelNonArgument for i in is_])

    # First check for conditions that are due to technical limitations of Maki and MerC and not
    # due to irreconcilable differences in macro and C semantics.

    # If any part of the type signature has a function pointer
    invocation_has_function_type = \
        any(i.IsExpansionTypeFunctionType or i.IsAnyArgumentTypeFunctionType for i in invocations)

    # TODO(Joey): Implement a way to translate these
    if invocation_has_function_type:
        logger.debug(f"Skipping {macro.Name} as it has a function pointer type")

    # If body contains a DeclRefExpr and is in a header file, skip
    # TODO(Joey/Brent): Find better way on Maki side to handle this
    invocation = next(iter(invocations))

    invocation_has_decl_ref_expr = invocation.DoesBodyContainDeclRefExpr
    if invocation_has_decl_ref_expr and invocation.DefinitionLocationFilename.endswith(".h"):
        logger.debug(f"Skipping {macro.Name} as it contains a DeclRefExpr")

    if (invocation_has_function_type or
        (invocation_has_decl_ref_expr and invocation.DefinitionLocationFilename.endswith(".h")):
        return False

    # Then check for semantic differences that would make replacing
    # this macro definition wiht C code incorrect.

    # ... Rest of original definition of ie_def()
    # (maybe we add logging statements for these checks as well)
```

**Review by PappasBrent — APPROVED** [review]

**Comment by PappasBrent** [comment]
Approving now, but before merging please either please either add a comment explaining why we need the `should_skip_macro()` function in addition to `ie_def()` (technical vs. theoretical limitations) or just move its checks into the definition of `ie_def()`.

File: macrotranslator.py
Comment (SilverMight) [reply macrotranslator.py]: For now I'm going to rename the function to clarify and consider revisiting our skip structure later (possibly returning a specific reason)

**Pull request merged**
